### ISSUE TITLE
Existing tools generate hexadecimal encoding, not base64

### ISSUE THREAD

**Issue by David263** [mdn/content#3850]
MDN URL: https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity

#### What information was incorrect, unhelpful, or incomplete?

The tool listed for generating a hashing for integrity attributes generates hexadecimal strings, not base64 as required by browsers. A search for similar tools turns up at least seven other tools, none of which appear to generate base64 output either.

The command lines given for generating a hashing appear to be valid for Linux, but it is unclear whether they would work in Mac or Windows computers.

#### Specific section or headline?

Tools for generating SRI hashes

#### What did you expect to see?

I expected to see working tools for generating such attributes.

#### Did you test this? If so, how?

I tried to test the recommended  online `SRI Hash Generator` at https://www.srihash.org, but found no way to generate a hashing without publishing my test case to the public Web, which I preferred not to do.

I reviewed the output of six hashing tools at site https://www.makeuseof.com/tag/free-hash-checkers-file-integrity/ .

I tested the `FileAlyzer 2` tool for Windows, which has configurable file hashing outputs.

I tested the `SHGW `tool (https://github.com/christopher-panayi/SHGW) on Windows.

<details>
<summary>MDN Content page report details</summary>

* Folder: `en-us/web/security/subresource_integrity`
* MDN URL: https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity
* GitHub URL: https://github.com/mdn/content/blob/main/files/en-us/web/security/subresource_integrity/index.html
* Last commit: https://github.com/mdn/content/commit/bba822dea342e78d30580205897c751bb354cbe3
* Document last modified: 2020-12-18T23:56:01.000Z

</details>

**Comment by David263** [comment]
Here is a simple .BAT or .CMD file for calculating the integrity value from a file in Windows. To use this, store this file once only in the SendTo folder (could be C:\Users\USER\AppData\Roaming\Microsoft\Windows\SendTo). Then right-click a file in the File Explorer, select `send to` and select the name you gave to the hashing tool file. You will see the integrity value in a command box. Select the value and right-click to copy it to the Clipboard. Press any key to dismiss the command box.

```
@echo off
REM set bits=256
set bits=384
openssl dgst -sha%bits% -binary %1% | openssl base64 -A > tmp
set /p a= < tmp
del tmp
echo sha%bits%-%a%
pause

```

**Comment by sideshowbarker** [comment]
> The tool listed for generating a hashing for integrity attributes generates hexadecimal strings, not base64 as required by browsers.

That’s not correct. In fact https://www.srihash.org/ does generate base64-encoded hashes.

For example, if I give it the url `https://github.githubassets.com/assets/behaviors-f8cc2f0b.js`, it gives me back this hash:

```
sha384-LAX6BB586rYlR5rNHsCBZtlxrf5B8fkf6UOkiHd6W1QPrZBwlA9+Vue1Z8q6NH6I
```
That’s not a hexadecimal string — it’s bash64-encoded.

> The command lines given for generating a hashing appear to be valid for Linux, but it is unclear whether they would work in Mac or Windows computers.

From the macOS terminal, the following command:
```bash
cat FILENAME.js | openssl dgst -sha384 -binary | openssl base64 -A
```
…definitely does work.

That same command should also work in any Windows environment which has the `cat` and `openssl` commands.

And in any Windows environment which has the `shasum`, `awk`, `xxd`, and `base64` tools installed, the following should work:
```bash
shasum -b -a 384 FILENAME.js | awk '{ print $1 }' | xxd -r -p | base64
```

> I expected to see working tools for generating such attributes.

The tools shown in the article work — both the online https://www.srihash.org/ tool and the command-line tools.

**Comment by sideshowbarker** [comment]
> Here is a simple .BAT or .CMD file for calculating the integrity value from a file in Windows. To use this, store this file once only in the SendTo folder (could be C:\Users\USER\AppData\Roaming\Microsoft\Windows\SendTo). Then right-click a file in the File Explorer, select `Send to...` and select the name you gave to the hashing tool file. You will see the integrity value in a command box. Select the value and right-click to copy it to the Clipboard. Press any key to dismiss the command box.
> 
> ```
> @echo off
> REM set bits=256
> set bits=384
> openssl dgst -sha%bits% -binary %1% | openssl base64 -A > tmp
> set /p a= < tmp
> del tmp
> echo sha%bits%-%a%
> pause
> ```

That looks worth adding to the article. It would be great if you could raise a pull a request to get it added.

**Comment by David263** [comment]
I am very sorry, but I have tried to use pull commands in the past and have gotten horribly lost in the workflow that follows. I have even tried using git tutorials to understand better, but have failed to understand them either. The full use of GitHub appears to be beyond my abilities to understand, perhaps because I am an old guy (75).

**Comment by sideshowbarker** [comment]
Yeah the workflow required to use command-line or desktop git tooling to edit articles is definitely a lot more complicated than the wiki-based system MDN had previously.

However, there’s an alternative that avoids the need to use command-line or desktop git and that seems no more complicated than the old wiki-based system. That is, you can edit any article directly online.

For example, if you open https://github.com/mdn/content/edit/main/files/en-us/web/security/subresource_integrity/index.html, you’ll get a form-based editing UI that allows you to edit the article source. And when you’re done with your edits, scroll to the bottom and push he **Commit changes** button. That will automatically create a pull request for you.

All that said, if you’d prefer I add the information from https://github.com/mdn/content/issues/3850#issuecomment-813519054 to the article myself, I’d be very happy to do that. It seems like it’d be very useful to developers working in Windows environments, so let’s definitely get it into the article — one way or another.

**Comment by David263** [comment]
I followed your directions to edit an article directly (just as a test, in this case). But I got an error message from GitHub saying I had to fork the repository (relying on my poor memory here). I did fork the repository, but the new instructions now say that I will have to submit my own pull request after I do my local edits. As I explained elsewhere, I am not smart enough to use pull requests.

Besides, I would rather have any of my edits reviewed by others first, even in cases where I am certain that my edits are correct. I have been wrong in many such cases before!

But, as you say elsewhere, I can simply submit an Issue and ask that someone else do the pull.

I hope that others who wish to edit don't get as confused as I did with the new procedures. But I do understand that there are several good reasons for them.
